### ISSUE TITLE
Using Dash through Hadoop (WASB)

### ISSUE THREAD

**Issue by denisgarci** [MicrosoftDX/Dash#3]
Hi,

Thank for this great project!

We're trying to connect our Hadoop cluster to Dash. We're having trouble configuring the `core-site.xml` for Dash. The Hadoop documentation doesn't specify how to configure a storage account with a custom domain name. We tried this for the `core-site.xml` key:
`fs.azure.account.key.{dash_account_name}.{custom_domain_name}`
and the account key as value.

When we try to access the Dash service through this WASB URI:
`wasb://{container_name}@{dash_account_name}.{custom_domain_name}/`
we get the following error:
`com.microsoft.azure.storage.StorageException: The server encountered an unknown failure:`.

Is there a way to specify a custom domain name for WASB in Hadoop? Any help would be appreciated!

**Comment by jamesbak** [comment]
Hi Denis,

Happy holidays! Sorry for the delay in responding...

Have you verified that the DASH service is working correctly using the custom domain name (ie. using a storage explorer or programmatically accessing via the custom domain url)? 

If the custom domain name is working correctly, then the change you made to `core-site.xml` is correct. The next thing to verify is to make sure the changes to `core-site.xml` are effective. Is your Hadoop cluster an HDInsight instance? If so, is it Windows or Linux? If you are using an Ambari managed cluster (eg. HDInsight on Linux), then you should make the config changes via the Ambari Web UI, rather than the `core-site.xml` file directly.

Next, ensure that the Azure storage client SDK JAR has been replaced on the cluster (the default SDK does not support automatic following of https redirections). If you are using a Linux cluster, SSH onto a head node and run the script located here: [https://github.com/MicrosoftDX/Dash/blob/master/Deployment/HDInsight/Linux/dash-linux-fixup-cache-files.sh](https://github.com/MicrosoftDX/Dash/blob/master/Deployment/HDInsight/Linux/dash-linux-fixup-cache-files.sh). This script downloads the JARs bundle, replaces the storage JAR & replaces the bundle.

That _should_ take care of everything. If not, ping back and we can iterate.

Cheers,
James

**Comment by denisgarci** [comment]
Hi James,

Thanks for you reply! We tried Dash using the Python package `azure-storage` and it was working. We are using a Hortonworks Linux cluster. Thanks for pointing out the modified Azure storage client JAR. We will try it out.
We have other questions:
* does Azure Storage Explorer work with Dash ?
* is Dash still under development?

Thanks again!

Cheers,
Denis

**Comment by jamesbak** [comment]
Hi Denis,

Happy New Year!

Good to know that your DASH installation is working. Although, we have never run a test directly against a HWK cluster, we have performed significant testing using a HDInsight Linux cluster (which is essentially the same thing). Therefore, we would anticipate that the process to enable DASH should be the same. Make sure you're using Ambari to add the extra settings for core-site. The script I pointed you at to update the Azure Storage JAR looks for all tar balls under `/hdp/apps` on the default filesystem for the cluster. If the HWK cluster sets up differently, you will need to modify the script accordingly.

Most third party storage explorers work with DASH (DASH supports a 'passthrough proxy' mode if it cannot determine that the client supports automatic HTTP redirections). Generally, the 1 hitch is getting the connection information entered. Specifically, Azure Storage Explorer accepts a custom endpoint URI on the connection string when attaching a new account, but falls back to the standard domain names when establishing a connection. This can be worked around by creating a `hosts` entry using the IP of your DASH endpoint and the standard pattern `accountname.blob.core.windows.net` as the DNS name. DASH will work in proxy mode meaning that it will not be ideal for high throughput (ie. > 10Gbps) usage, but this is not the normal pattern for storage explorers anyway. 

I would highly recommend the **Azure Management Studio** tool from Cerebrata [http://www.cerebrata.com/products/azure-management-studio/introduction](http://www.cerebrata.com/products/azure-management-studio/introduction). You still need to do the `hosts` trick to connect, but it is an excellent tool for managing storage and other Azure resources.

DASH is still very much under development. We did have a slow second half of 2016, but we expect to deliver a new v2.0 release early in 2017 with a significant new set of features, including geo-distribution. Stay tuned!

Cheers,
James
